recommenders/models/sasrec/util.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import sys
+import copy
+import random
+import numpy as np
+import tensorflow as tf
+from collections import defaultdict
+from tqdm import tqdm
+import time
+
+
+def data_partition(fname):
+    usernum = 0
+    itemnum = 0
+    User = defaultdict(list)
+    user_train = {}
+    user_valid = {}
+    user_test = {}
+    # assume user/item index starting from 1
+    # f = open('data/%s.txt' % fname, 'r')
+    f = open(fname, "r")
+    for line in f:
+        u, i = line.rstrip().split(" ")
+        u = int(u)
+        i = int(i)
+        usernum = max(u, usernum)
+        itemnum = max(i, itemnum)
+        User[u].append(i)
+
+    for user in User:
+        nfeedback = len(User[user])
+        if nfeedback < 3:
+            user_train[user] = User[user]
+            user_valid[user] = []
+            user_test[user] = []
+        else:
+            user_train[user] = User[user][:-2]
+            user_valid[user] = []
+            user_valid[user].append(User[user][-2])
+            user_test[user] = []
+            user_test[user].append(User[user][-1])
+    return [user_train, user_valid, user_test, usernum, itemnum]
```

Comment:
I think when a method returns too many items, it's a good sign of using an object instead. I would recommend turning the code into a class to avoid return a list of many items.